rootfs/api/models/app.py

```diff
@@ -0,0 +1,528 @@
+    id = db_models.SlugField(max_length=24, unique=True, default=select_app_name,
+                             validators=validators)
+    structure = JSONField(default={}, blank=True, validators=[validate_app_structure])
+
+    class Meta:
+        permissions = (('use_app', 'Can use app'),)
+        db_table = 'app'
+        app_label = 'api'
+
+    @property
+    def _scheduler(self):
+        mod = importlib.import_module(settings.SCHEDULER_MODULE)
+        return mod.SchedulerClient(settings.SCHEDULER_URL,
+                                   settings.SCHEDULER_AUTH,
+                                   settings.SCHEDULER_OPTIONS)
+
+    def __str__(self):
+        return self.id
+
+    @property
+    def url(self):
+        return self.id + '.' + settings.DEIS_DOMAIN
+
+    def _get_job_id(self, container_type):
+        app = self.id
+        release = self.release_set.latest()
+        version = "v{}".format(release.version)
+        job_id = "{app}_{version}.{container_type}".format(**locals())
+        return job_id
+
+    def _get_command(self, container_type):
+        try:
+            # if this is not procfile-based app, ensure they cannot break out
+            # and run arbitrary commands on the host
+            # FIXME: remove slugrunner's hardcoded entrypoint
+            release = self.release_set.latest()
+            if release.build.dockerfile or not release.build.sha:
+                return "bash -c '{}'".format(release.build.procfile[container_type])
+
+            return 'start {}'.format(container_type)
+        # if the key is not present or if a parent attribute is None
+        except (KeyError, TypeError, AttributeError):
+            # handle special case for Dockerfile deployments
+            return '' if container_type == 'cmd' else 'start {}'.format(container_type)
+
+    def log(self, message, level=logging.INFO):
+        """Logs a message in the context of this application.
+
+        This prefixes log messages with an application "tag" that the customized deis-logspout will
+        be on the lookout for.  When it's seen, the message-- usually an application event of some
+        sort like releasing or scaling, will be considered as "belonging" to the application
+        instead of the controller and will be handled accordingly.
+        """
+        logger.log(level, "[{}]: {}".format(self.id, message))
+
+    def create(self, *args, **kwargs):
+        """Create a new application with an initial config and release"""
+        config = Config.objects.create(owner=self.owner, app=self)
+        Release.objects.create(version=1, owner=self.owner, app=self, config=config, build=None)
+
+    def delete(self, *args, **kwargs):
+        """Delete this application including all containers"""
+        try:
+            # attempt to remove containers from the scheduler
+            self._destroy_containers([c for c in self.container_set.exclude(type='run')])
+        except RuntimeError:
+            pass
```

Comment:
at least log here?

again, don't know what the old code did so not a blocker if this is copied